Incomplete remote response handler after transport close in integration tests

### CI Link

https://gradle-enterprise.elastic.co/s/ma2rjwewae6hi

### Repro line

./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.action.termvectors.MultiTermVectorsIT" -Dtests.seed=D2D36E011E6F103B -Dtests.locale=en-US -Dtests.timezone=UTC -Druntime.java=17

### Does it reproduce?

No

### Applicable branches

master

### Failure history

_No response_

### Failure excerpt

I don't think the failed test suite (`MultiTermVectorsIT`) has anything to do with this failure.

```
Apr 28, 2022 7:07:07 AM com.carrotsearch.randomizedtesting.ThreadLeakControl checkThreadLeaks |  
-- | --
  | 2> SEVERE: 73 threads leaked from SUITE scope at org.elasticsearch.action.termvectors.MultiTermVectorsIT: |  
  | 2>    1) Thread[id=308, name=elasticsearch[node_s3][generic][T#1], state=WAITING, group=TGRP-MultiTermVectorsIT] |  
  | 2>         at java.base@17.0.2/jdk.internal.misc.Unsafe.park(Native Method) |  
  | 2>         at java.base@17.0.2/java.util.concurrent.locks.LockSupport.park(LockSupport.java:341) |  
  | 2>         at java.base@17.0.2/java.util.concurrent.LinkedTransferQueue$Node.block(LinkedTransferQueue.java:470) |  
  | 2>         at java.base@17.0.2/java.util.concurrent.ForkJoinPool.unmanagedBlock(ForkJoinPool.java:3463) |  
  | 2>         at java.base@17.0.2/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3434) |  
  | 2>         at java.base@17.0.2/java.util.concurrent.LinkedTransferQueue.awaitMatch(LinkedTransferQueue.java:669) |  
  | 2>         at java.base@17.0.2/java.util.concurrent.LinkedTransferQueue.xfer(LinkedTransferQueue.java:616) |  
  | 2>         at java.base@17.0.2/java.util.concurrent.LinkedTransferQueue.take(LinkedTransferQueue.java:1286) |  
  | 2>         at java.base@17.0.2/java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1062) |  
  | 2>         at java.base@17.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1122) |  
  | 2>         at java.base@17.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) |  
  | 2>         at java.base@17.0.2/java.lang.Thread.run(Thread.java:833)
```

And anther relevant snippet:

```
java.lang.RuntimeException: file handle leaks: [FileChannel(/dev/shm/elastic+elasticsearch+main+intake+multijob+part1/server/build/testrun/internalClusterTest/temp/org.elasticsearch.action.termvectors.MultiTermVectorsIT_D2D36E011E6F103B-001/tempDir-002/node_s2/_state/write.lock), FileChannel(/dev/shm/elastic+elasticsearch+main+intake+multijob+part1/server/build/testrun/internalClusterTest/temp/org.elasticsearch.action.termvectors.MultiTermVectorsIT_D2D36E011E6F103B-001/tempDir-002/node_s0/_state/write.lock), FileChannel(/dev/shm/elastic+elasticsearch+main+intake+multijob+part1/server/build/testrun/internalClusterTest/temp/org.elasticsearch.action.termvectors.MultiTermVectorsIT_D2D36E011E6F103B-001/tempDir-002/node_s1/node.lock), FileChannel(/dev/shm/elastic+elasticsearch+main+intake+multijob+part1/server/build/testrun/internalClusterTest/temp/org.elasticsearch.action.termvectors.MultiTermVectorsIT_D2D36E011E6F103B-001/tempDir-002/node_s2/_state/_m.cfs), FileChannel(/dev/shm/elastic+elasticsearch+main+intake+multijob+part1/server/build/testrun/internalClusterTest/temp/org.elasticsearch.action.termvectors.MultiTermVectorsIT_D2D36E011E6F103B-001/tempDir-002/node_s1/_state/write.lock), FileChannel(/dev/shm/elastic+elasticsearch+main+intake+multijob+part1/server/build/testrun/internalClusterTest/temp/org.elasticsearch.action.termvectors.MultiTermVectorsIT_D2D36E011E6F103B-001/tempDir-002/node_s3/node.lock), FileChannel(/dev/shm/elastic+elasticsearch+main+intake+multijob+part1/server/build/testrun/internalClusterTest/temp/org.elasticsearch.action.termvectors.MultiTermVectorsIT_D2D36E011E6F103B-001/tempDir-002/node_s2/node.lock), FileChannel(/dev/shm/elastic+elasticsearch+main+intake+multijob+part1/server/build/testrun/internalClusterTest/temp/org.elasticsearch.action.termvectors.MultiTermVectorsIT_D2D36E011E6F103B-001/tempDir-002/node_s0/node.lock)] |  
-- | --
  | at org.apache.lucene.tests.mockfile.LeakFS.onClose(LeakFS.java:63) |  
  | at org.apache.lucene.tests.mockfile.FilterFileSystem.close(FilterFileSystem.java:70) |  
  | at org.apache.lucene.tests.mockfile.FilterFileSystem.close(FilterFileSystem.java:71) |  
  | at org.apache.lucene.tests.util.TestRuleTemporaryFilesCleanup.afterAlways(TestRuleTemporaryFilesCleanup.java:223) |  
  | at com.carrotsearch.randomizedtesting.rules.TestRuleAdapter$1.afterAlways(TestRuleAdapter.java:31) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:43) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53) |  
  | at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43) |  
  | at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44) |  
  | at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60) |  
  | at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47) |  
  | at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) |  
  | at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375) |  
  | at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831) |  
  | at java.base/java.lang.Thread.run(Thread.java:833) |  
  |   |  
  | Caused by: |  
  | java.lang.Exception |  
  | at org.apache.lucene.tests.mockfile.LeakFS.onOpen(LeakFS.java:46) |  
  | at org.apache.lucene.tests.mockfile.HandleTrackingFS.callOpenHook(HandleTrackingFS.java:82) |  
  | at org.apache.lucene.tests.mockfile.HandleTrackingFS.newFileChannel(HandleTrackingFS.java:202) |  
  | at org.apache.lucene.tests.mockfile.HandleTrackingFS.newFileChannel(HandleTrackingFS.java:171) |  
  | at java.base/java.nio.channels.FileChannel.open(FileChannel.java:298) |  
  | at java.base/java.nio.channels.FileChannel.open(FileChannel.java:357) |  
  | at org.apache.lucene.store.NativeFSLockFactory.obtainFSLock(NativeFSLockFactory.java:112) |  
  | at org.apache.lucene.store.FSLockFactory.obtainLock(FSLockFactory.java:43) |  
  | at org.apache.lucene.store.BaseDirectory.obtainLock(BaseDirectory.java:44) |  
  | at org.apache.lucene.index.IndexWriter.<init>(IndexWriter.java:947) |  
  | at org.elasticsearch.gateway.PersistedClusterStateService.createIndexWriter(PersistedClusterStateService.java:254) |  
  | at org.elasticsearch.gateway.PersistedClusterStateService.createWriter(PersistedClusterStateService.java:216) |  
  | at org.elasticsearch.gateway.GatewayMetaState$LucenePersistedState.<init>(GatewayMetaState.java:448) |  
  | at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:131) |  
  | at org.elasticsearch.node.Node.start(Node.java:1167) |  
  | at org.elasticsearch.test.InternalTestCluster$NodeAndClient.startNode(InternalTestCluster.java:982) |  
  | at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) |  
  | at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) |  
  | at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:714) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) |  
  | ... 1 more
```

Link to logs: https://gradle-enterprise.elastic.co/s/ma2rjwewae6hi/console-log#L9010

The actual failure is here:

```
org.elasticsearch.action.termvectors.MultiTermVectorsIT > classMethod FAILED
    java.lang.AssertionError: {127.0.0.1:39821}{GsBDsf__T_-ZYuus_____w}{127.0.0.1}{127.0.0.1:39821} vs {node_s0}{oy5xD2ePQaeEfHPovqb57w}{m4jRNYcNSSy_G0iawFWoLA}{127.0.0.1}{127.0.0.1:46787}{cdfhilmrstw}
        at __randomizedtesting.SeedInfo.seed([D2D36E011E6F103B]:0)
        at org.elasticsearch.transport.TransportService.doStop(TransportService.java:305)
        at org.elasticsearch.common.component.AbstractLifecycleComponent.stop(AbstractLifecycleComponent.java:68)
        at org.elasticsearch.node.Node.stop(Node.java:1301)
        at org.elasticsearch.node.Node.close(Node.java:1320)
        at org.elasticsearch.test.InternalTestCluster$NodeAndClient.close(InternalTestCluster.java:1054)
        at org.elasticsearch.core.IOUtils.close(IOUtils.java:73)
        at org.elasticsearch.core.IOUtils.close(IOUtils.java:115)
        at org.elasticsearch.core.IOUtils.close(IOUtils.java:98)
        at org.elasticsearch.test.InternalTestCluster.close(InternalTestCluster.java:905)
        at org.elasticsearch.core.IOUtils.close(IOUtils.java:73)
        at org.elasticsearch.core.IOUtils.close(IOUtils.java:115)
        at org.elasticsearch.core.IOUtils.close(IOUtils.java:98)
        at org.elasticsearch.test.ESIntegTestCase.clearClusters(ESIntegTestCase.java:530)
        at org.elasticsearch.test.ESIntegTestCase.afterClass(ESIntegTestCase.java:2222)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:909)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
        at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
        at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

Throwing an `AssertionError` here means we don't complete the rest of the pending transport handlers which can indeed leak a file handle, but that's just noise. I've updated the title.

--- a/docs/changelog/86315.yaml
+++ b/docs/changelog/86315.yaml
@@ -1,0 +1,6 @@
+pr: 86315
+summary: Reject `openConnection` attempt while closing
+area: Network
+type: bug
+issues:
+ - 86249

--- a/server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java
+++ b/server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java
@@ -75,7 +75,20 @@ public class ClusterConnectionManager implements ConnectionManager {
     @Override
     public void openConnection(DiscoveryNode node, ConnectionProfile connectionProfile, ActionListener<Transport.Connection> listener) {
         ConnectionProfile resolvedProfile = ConnectionProfile.resolveConnectionProfile(connectionProfile, defaultProfile);
-        internalOpenConnection(node, resolvedProfile, listener);
+        if (connectingRefCounter.tryIncRef()) {
+            var success = false;
+            final var release = new RunOnce(connectingRefCounter::decRef);
+            try {
+                internalOpenConnection(node, resolvedProfile, ActionListener.runBefore(listener, release::run));
+                success = true;
+            } finally {
+                if (success == false) {
+                    release.run();
+                }
+            }
+        } else {
+            listener.onFailure(new ConnectTransportException(node, "connection manager is closed"));
+        }
     }
 
     /**
@@ -124,7 +137,7 @@ public class ClusterConnectionManager implements ConnectionManager {
         }
 
         if (connectingRefCounter.tryIncRef() == false) {
-            listener.onFailure(new IllegalStateException("connection manager is closed"));
+            listener.onFailure(new ConnectTransportException(node, "connection manager is closed"));
             return;
         }
 

--- a/server/src/test/java/org/elasticsearch/transport/ClusterConnectionManagerTests.java
+++ b/server/src/test/java/org/elasticsearch/transport/ClusterConnectionManagerTests.java
@@ -39,6 +39,7 @@ import java.util.concurrent.BrokenBarrierException;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Semaphore;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -46,6 +47,7 @@ import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
 
 import static org.elasticsearch.test.ActionListenerUtils.anyActionListener;
+import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.mockito.ArgumentMatchers.any;
@@ -488,6 +490,27 @@ public class ClusterConnectionManagerTests extends ESTestCase {
         assertEquals(0, connectionManager.size());
         assertEquals(0, nodeConnectedCount.get());
         assertEquals(0, nodeDisconnectedCount.get());
+    }
+
+    public void testConnectAfterClose() {
+        connectionManager.close();
+        final var node = new DiscoveryNode("", new TransportAddress(InetAddress.getLoopbackAddress(), 0), Version.CURRENT);
+
+        final var openConnectionFuture = new PlainActionFuture<Transport.Connection>();
+        connectionManager.openConnection(node, connectionProfile, openConnectionFuture);
+        assertTrue(openConnectionFuture.isDone());
+        assertThat(
+            expectThrows(ExecutionException.class, ConnectTransportException.class, openConnectionFuture::get).getMessage(),
+            containsString("connection manager is closed")
+        );
+
+        final var connectToNodeFuture = new PlainActionFuture<Releasable>();
+        connectionManager.connectToNode(node, connectionProfile, (c, p, l) -> fail("should not be called"), connectToNodeFuture);
+        assertTrue(connectToNodeFuture.isDone());
+        assertThat(
+            expectThrows(ExecutionException.class, ConnectTransportException.class, connectToNodeFuture::get).getMessage(),
+            containsString("connection manager is closed")
+        );
     }
 
     private static class TestConnect extends CloseableConnection {

